### PR TITLE
Feature/add new group permissions

### PR DESCRIPTION
**Pull request by KaterinaWire**
## What's new in this PR?

Add new permission logic for group conversations.

### CODE REVIEW

**Comment by zenkins** [comment]
Can one of the admins verify this patch?

File: Source/Model/User/ZMUser+Permissions.swift [hunk 1]
```diff
@@ -41,18 +67,40 @@ public extension ZMUser {
...
+            guard conversation.teamRemoteIdentifier == nil || !isGuest(in: conversation), conversation.isSelfAnActiveMember else { return false }
+            return permissions?.contains(.addRemoveConversationMember) ?? true
+        }
     }
     
     @objc(canRemoveUserFromConversation:)
     func canRemoveUser(from conversation: ZMConversation) -> Bool {
-        return canAddUser(to: conversation)
+        if conversation.conversationType == .group {
+            let canModifyConversationName = self.participantRoles.filter({$0.conversation == conversation}).first?.role?.actions.contains(where: {$0.name == ActionNames.removeConversationMember.name})
```
Comment (marcoconti83) [inline Source/Model/User/ZMUser+Permissions.swift]: I would extract the following:
```
self.participantRoles.filter({$0.conversation == conversation}).first?.role?.actions.contains(where: {$0.name == xxx})
```

To a method, something like `ZMUser.hasRoleWithAction(action: ActionNames, conversation: ZMConversation)`

File: Source/Model/User/ZMUser+Permissions.swift [hunk 2]
```diff
@@ -41,18 +67,40 @@ public extension ZMUser {
     
     @objc(canAddUserToConversation:)
     func canAddUser(to conversation: ZMConversation) -> Bool {
-        guard conversation.teamRemoteIdentifier == nil || !isGuest(in: conversation), conversation.isSelfAnActiveMember else { return false }
-        return permissions?.contains(.addRemoveConversationMember) ?? true
+        if conversation.conversationType == .group {
+            let canModifyConversationName = self.participantRoles.filter({$0.conversation == conversation}).first?.role?.actions.contains(where: {$0.name == ActionNames.addConversationMember.name})
```
Comment (billypchan) [inline Source/Model/User/ZMUser+Permissions.swift]: We can create a helper method to search for a action name of a user, e.g.
```

File: Tests/Source/Model/User/ZMUserTests+Permissions.swift [hunk 3]
```diff
@@ -509,4 +511,90 @@ class ZMUserTests_Permissions: ModelObjectsTests {
         // then
         XCTAssertFalse(ZMUser.selfUser(in: uiMOC).canModifyTitle(in: conversation))
     }
+    
+    func testThatConversationTitleCanBeModified_ByGroupParticipant() {
```
Comment (marcoconti83) [inline Tests/Source/Model/User/ZMUserTests+Permissions.swift]: I would also add opposite tests, tests that the user can not modify conversation if the role action is not there

File: Source/Model/User/ZMUser+Permissions.swift [hunk 4]
```diff
@@ -18,6 +18,32 @@
 
 import Foundation
 
+enum ActionNames {
```
Comment (billypchan) [inline Source/Model/User/ZMUser+Permissions.swift]: I would suggest use a singlar from `ConversationAction`

Comment (KaterinaWire) [reply Source/Model/User/ZMUser+Permissions.swift]: Fixed

File: Source/Model/User/ZMUser+Permissions.swift (hunk 2 above)
Comment (KaterinaWire) [reply Source/Model/User/ZMUser+Permissions.swift]: Added a helper method `hasRoleWithAction`

**Pull request merged**